core/lib/contexts.ts

```diff
@@ -0,0 +1,445 @@
+    let partialEvaluate = (primitive) => {
+        let result = evaluatePrimitive(primitive, event, eventType, eventSchema);
+        if (result && result.length !== 0) {
+            return result;
+        }
+    };
+    let generatedContexts = map(normalizedInputs, partialEvaluate);
+    return [].concat(...compact(generatedContexts));
+}
+
+export function evaluatePrimitive(contextPrimitive: ContextPrimitive,
+                           event: SelfDescribingJson,
+                           eventType: string,
+                           eventSchema: string) : Array<SelfDescribingJson> | undefined {
+    if (isSelfDescribingJson(contextPrimitive)) {
+        return [contextPrimitive as SelfDescribingJson];
+    } else if (isContextGenerator(contextPrimitive)) {
+        let generatorOutput = buildGenerator(contextPrimitive as ContextGenerator, event, eventType, eventSchema);
+        if (isSelfDescribingJson(generatorOutput)) {
+            return [generatorOutput as SelfDescribingJson];
+        } else if (Array.isArray(generatorOutput)) {
+            return generatorOutput;
+        }
+    }
+    return undefined;
+}
+
+export function evaluateProvider(provider: ConditionalContextProvider,
+                          event: SelfDescribingJson,
+                          eventType: string,
+                          eventSchema: string): Array<SelfDescribingJson> {
+    if (isFilterProvider(provider)) {
+        let filter : ContextFilter = (provider as FilterProvider)[0];
+        let filterResult = false;
+        try {
+            let args = {
+                event: event,
+                eventType: eventType,
+                eventSchema: eventSchema
+            };
+            filterResult = filter(args);
+        } catch(error) {
+            filterResult = false;
+        }
+        if (filterResult === true) {
+            return generatePrimitives((provider as FilterProvider)[1], event, eventType, eventSchema);
+        }
+    } else if (isRuleSetProvider(provider)) {
+        if (matchSchemaAgainstRuleSet((provider as RuleSetProvider)[0], eventSchema)) {
+            return generatePrimitives((provider as RuleSetProvider)[1], event, eventType, eventSchema);
+        }
+    }
+    return [];
+}
+
+export function generateConditionals(providers: Array<ConditionalContextProvider> | ConditionalContextProvider,
+                              event: SelfDescribingJson,
+                              eventType: string,
+                              eventSchema: string) : Array<SelfDescribingJson> {
+    let normalizedInput : Array<ConditionalContextProvider> = normalizeToArray(providers);
+    let partialEvaluate = (provider) => {
+        let result = evaluateProvider(provider, event, eventType, eventSchema);
+        if (result && result.length !== 0) {
+            return result;
+        }
+    };
+    let generatedContexts = map(normalizedInput, partialEvaluate);
+    return [].concat(...compact(generatedContexts));
+}
```

Comment:
can't we generalize parts of `generate{Conditionals/Primitives}`? They seems to share a lot of logic